### ISSUE TITLE
CRD podTemplate schema forces to declare an empty spec.containers

### ISSUE THREAD

**Issue by thbkrkr** [elastic/cloud-on-k8s#1822]
By working to bump the CRD version (#1782 ), TestSmoke fails with: `spec.podTemplate.spec.containers in body must be of type array: "null"`.

With the new controller-gen that comes with kubebuilder 2, the `podTemplate` object schema is fully inlined in our CRDs.

As there must be at least one container in a Pod, the `Containers []Container 'json:"containers"'` does not have the `omitempty` tag.

So when a `podTemplate.spec` property is defined like a `podTemplate.spec.securityContext`, then the `podTemplate.spec.containers` is set by default to `nil` and this is what causes the validation error.

To pass validation, `podTemplate.spec.containers` should always be set to an empty array if not used. It seems very compelling to always have to ask for this declaration.

Example:

```yaml
apiVersion: kibana.k8s.elastic.co/v1beta1
kind: Kibana
metadata:
  name: kibana-sample
spec:
  version: 7.3.0
  nodeCount: 1
  elasticsearchRef:
    name: "elasticsearch-sample"
  podTemplate:
    spec:
      updateStrategy:
        changeBudget:
          maxSurge: 1
          maxUnavailable: 0
      containers: []  # <-------
```

We have identified 4 potential solutions to solve this problem.

#### 1) Use a mutating admission webhook

That we do not have currently.

#### 2) Remove the `podTemplate` schema from the CRD

We loose the `podTemplate` validation which is a very nice feature to have.

#### 3) Have our own `podSpec` struct (copied from the original)

This implies maintaining a kind of fork.

#### 4) Make the `podTemplate.spec.containers` field optional

This involves applying some post-treatment after the CRD generation to remove the `containers` item from the `podTemplate.spec.required` string array.

**Comment by thbkrkr** [comment]
Option 4) does not work as expected.
The operator fails at runtime:

```
2019-10-04T10:37:45.147+0200    ERROR   controller-runtime.controller   Reconciler error        {"ver": "0.10.0-SNAPSHOT-25caaf10", "controller": "kibana-association-controller", "re
quest": "default/kibana-sample", "error": "Kibana.kibana.k8s.elastic.co \"kibana-sample\" is invalid: []: Invalid value: map[string]interface {}{\"metadata\":map[string]interface {}{
\"generation\":2, \"creationTimestamp\":\"2019-10-04T08:37:43Z\", \"selfLink\":\"/apis/kibana.k8s.elastic.co/v1beta1/namespaces/default/kibanas/kibana-sample\", \"uid\":\"3c1451a9-e6
82-11e9-b029-42010a840016\", \"resourceVersion\":\"2555031\", \"annotations\":map[string]interface {}{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"k
ibana.k8s.elastic.co/v1beta1\\\",\\\"kind\\\":\\\"Kibana\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"name\\\":\\\"kibana-sample\\\",\\\"namespace\\\":\\\"default\\\"},\\\"spec\\
\":{\\\"elasticsearch\\\":{\\\"auth\\\":{\\\"secret\\\":{\\\"key\\\":\\\"user\\\",\\\"name\\\":\\\"secret-name\\\"}},\\\"certificateAuthorities\\\":{\\\"secretName\\\":\\\"my-ca-cert
\\\"},\\\"url\\\":\\\"https://url.to.elasticsearch:9200\\\"},\\\"nodeCount\\\":1,\\\"podTemplate\\\":{\\\"spec\\\":{\\\"updateStrategy\\\":{\\\"changeBudget\\\":null,\\\"maxSurge\\\"
:1,\\\"maxUnavailable\\\":0}}},\\\"version\\\":\\\"7.4.0\\\"}}\\n\"}, \"finalizers\":[]interface {}{\"finalizer.association.kibana.k8s.elastic.co/elasticsearch\", \"finalizer.associa
tion.kibana.k8s.elastic.co/external-user\"}, \"namespace\":\"default\", \"name\":\"kibana-sample\"}, \"spec\":map[string]interface {}{\"http\":map[string]interface {}{\"service\":map
[string]interface {}{\"metadata\":map[string]interface {}{\"creationTimestamp\":interface {}(nil)}, \"spec\":map[string]interface {}{}}, \"tls\":map[string]interface {}{\"certificate
\":map[string]interface {}{}}}, \"podTemplate\":map[string]interface {}{\"metadata\":map[string]interface {}{\"creationTimestamp\":interface {}(nil)}, \"spec\":map[string]interface {
}{\"containers\":interface {}(nil)}}, \"version\":\"7.4.0\", \"nodeCount\":1, \"elasticsearchRef\":map[string]interface {}{\"name\":\"\"}}, \"kind\":\"Kibana\", \"apiVersion\":\"kiba
na.k8s.elastic.co/v1beta1\"}: validation failure list:\nspec.podTemplate.spec.containers in body must be of type array: \"null\""}
github.com/go-logr/zapr.(*zapLogger).Error                                                                                                                                                    /home/thb/dev/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128                                                                                                             
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler                                                                                                         /home/thb/dev/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:218                                                                         sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem                                                                                                      /home/thb/dev/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:192                                                                         sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker                                                                                                                   /home/thb/dev/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:171                                                                         k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1                                                                                                                                           /home/thb/dev/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152                                                                        k8s.io/apimachinery/pkg/util/wait.JitterUntil                                                                                                                                                 /home/thb/dev/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153                                                                        k8s.io/apimachinery/pkg/util/wait.Until                                                                                                                                                       /home/thb/dev/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88
```

**Comment by thbkrkr** [comment]
Option 2) was implemented by #1782. Then #1852 revisited this to use kustomize for the `podTemplate` removal.
